sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2WriteRequirements.scala

```diff
@@ -0,0 +1,107 @@
+import org.apache.spark.sql.connector.write.RequiresDistributionAndOrdering
+import org.apache.spark.sql.internal.SQLConf
+
+// this rule does resolution in the optimizer because some nodes like OverwriteByExpression
+// must undergo the expression optimization before we can construct a logical write
+object V2WriteRequirements extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transformDown {
+    case V2BatchWriteCommand(write: RequiresDistributionAndOrdering, query, false) =>
+      val sqlConf = SQLConf.get
+      val resolver = sqlConf.resolver
+
+      val distribution = write.requiredDistribution match {
+        case d: OrderedDistribution =>
+          d.ordering.map(e => toCatalyst(e, query, resolver))
+        case d: ClusteredDistribution =>
+          d.clustering.map(e => toCatalyst(e, query, resolver))
+        case _: UnspecifiedDistribution =>
+          Array.empty[catalyst.expressions.Expression]
+      }
+
+      val queryWithDistribution = if (distribution.nonEmpty) {
+        val numShufflePartitions = sqlConf.numShufflePartitions
+        // repartition by expression will correctly pick range or hash partitioning
+        // based on whether we have an ordered or clustered distribution
+        // all expressions in OrderedDistribution will be mapped into SortOrder
+        // all expressions in ClusteredDistribution will be mapped into Expression
+        RepartitionByExpression(distribution, query, numShufflePartitions)
+      } else {
+        query
+      }
+
+      val ordering = write.requiredOrdering.toSeq
+        .map(e => toCatalyst(e, query, resolver))
+        .asInstanceOf[Seq[catalyst.expressions.SortOrder]]
+
+      val queryWithDistributionAndOrdering = if (ordering.nonEmpty) {
+        Sort(ordering, global = false, queryWithDistribution)
+      } else {
+        queryWithDistribution
+      }
+
+      // set aligned flag to true to make the rule idempotent
+      V2BatchWriteCommand(write, queryWithDistributionAndOrdering, aligned = true)
+  }
+
+  private def toCatalyst(
+      expr: Expression,
+      query: LogicalPlan,
+      resolver: Resolver): catalyst.expressions.Expression = {
+    expr match {
+      case SortValue(child, direction, nullOrdering) =>
+        val catalystChild = toCatalyst(child, query, resolver)
+        SortOrder(catalystChild, toCatalyst(direction), toCatalyst(nullOrdering), Set.empty)
+      case ref: FieldReference =>
+        // this part is controversial as we perform resolution in the optimizer
+        // we cannot perform this step in the analyzer since we need to optimize expressions
+        // in nodes like OverwriteByExpression before constructing a logical write
+        query.resolve(ref.parts, resolver) match {
+          case Some(attr) => attr
+          case None => throw new AnalysisException(s"Cannot resolve '$ref' using ${query.output}")
+        }
+      case _ =>
```

Comment:
I think it would be helpful to support literals and transforms that are built-in Spark functions as well. For example, a timestamp column might use `date_trunc('HOUR', ts)` to co-locate data for a given hour. This doesn't need to happen in the initial commit, but transforms will be needed fairly early on.

Reply:
I agree.